### PR TITLE
Add upload data styles to Splunk uploader

### PR DESCRIPTION
**Pull request by Tatsuya-hasegawa**
Hi

I have added the post styles to Splunk by Splunk Uploader.
It's very useful for msticpy Splunk uploader's users to post data by json or text style.

ASIS, all the data frame are posted with the csv style by using to_csv() method.
It occurs event fragments by depending on the data to upload.

So, I developed a switch by using "sourcetype" argument. The main change is below. 

```python
__post_data
        for row in data.iterrows():
            if sourcetype=="csv":
                data = row[1].to_csv()  # type: ignore
            elif sourcetype=="json":
                data = row[1].to_json()  # type: ignore
            else:
                data = row[1].to_string()  # type: ignore
```

In addition, I renamed table_name to sourcetype at many locations.
Because, table_name is not usual word in Splunk.
On the other hand, pylint was failed in upload_file and upload_folder methods due to W0237 so I remain table_name in only these methods.

pylint...................................................................Failed
msticpy/data/uploaders/splunk_uploader.py:166:4: W0237: Parameter 'table_name' has been renamed to 'sourcetype' in overriding 'SplunkUploader.upload_file' method (arguments-renamed)
msticpy/data/uploaders/splunk_uploader.py:218:4: W0237: Parameter 'table_name' has been renamed to 'sourcetype' in overriding 'SplunkUploader.upload_folder' method (arguments-renamed)

Finally, my code commit has been passed by the commit test.

```shell
(base) hacket@hackeTlab msticpy % git commit -m "add_post_data_styles_to_splunk_uploader"
Check Yaml...........................................(no files to check)Skipped
Check JSON...........................................(no files to check)Skipped
Trim Trailing Whitespace.................................................Passed
black....................................................................Passed
pylint...................................................................Passed
flake8...................................................................Passed
isort (python)...........................................................Passed
pydocstyle...............................................................Passed
check_reqs_all...........................................................Passed
[mod_splunk_uploader_post 10646520] add_post_data_styles_to_splunk_uploader
 2 files changed, 41 insertions(+), 18 deletions(-)
 ```
 
 Best regards,

### CODE REVIEW

**Comment by Tatsuya-hasegawa** [comment]
My Splunk upload test is like this.

image

## csv example for the test data is bad!
image


## json example for the test data is great!
image

## text example for the test data is not bad!
image

Thanks in advance.

File: msticpy/data/uploaders/splunk_uploader.py [hunk 1]
```diff
@@ -199,7 +210,7 @@ def upload_file(  # type: ignore
             table_name = path.stem
         self._post_data(
             data=data,
-            table_name=table_name,
+            sourcetype=table_name,
```
Comment (ianhelle) [inline msticpy/data/uploaders/splunk_uploader.py:213]: Should this be sourcetype? like the other functions?

File: msticpy/data/uploaders/splunk_uploader.py [hunk 2]
```diff
@@ -116,7 +123,7 @@ def _post_data(
     def upload_df(  # type: ignore
         self,
         data: pd.DataFrame,
-        table_name: Optional[str],
+        sourcetype: Optional[str],
```
Comment (ianhelle) [inline msticpy/data/uploaders/splunk_uploader.py:126]: given that you are changing the parameter name, I would try to keep backward compatibility and maybe opt for json as the default. I have a mild preference for snake-case compound words like this - source_type vs sourcetype. I'm not super concerned about that though
```python

    source_type = kwargs.get("source_type", kwargs.get("table_name", "json"))
```

File: msticpy/data/uploaders/splunk_uploader.py [hunk 3]
```diff
@@ -252,14 +265,19 @@ def upload_folder(  # type: ignore
...
                 data=data,
-                table_name=table_name,
+                sourcetype=table_name,
                 index_name=index_name,
                 host=host,
                 create_index=create_index,
             )
             f_progress.update(1)
             if self._debug is True:
-                print(f"{str(path)} uploaded to {table_name}")
```
Comment (ianhelle) [inline msticpy/data/uploaders/splunk_uploader.py:262]: probably get rid of references to table name in the body of the function - if it isn't actually the table_name, it's bit confusing.

**Review by ianhelle — CHANGES_REQUESTED** [review]
Looks great - had a couple of minor comments..

File: docs/source/data_acquisition/UploadData.rst [hunk 4]
```diff
@@ -104,24 +104,27 @@ Uploading a DataFrame to Splunk
...
 
 Uploading a File to Splunk
 ^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 To upload a file to Splunk pass the path to the file to ``.upload_file()`` along with the name of the index you
-want the data uploaded to. By default a comma separated value file is expected but if you have some other separator
-value you can pass this with the ``delim`` parameter. You can specify a table name to upload the data to with that ``table_name``
-parameter but by default the uploader will upload to a table with the same name as the file. As with uploading a DataFrame
+want the data uploaded to. By default a comma separated value file is expected but if your file has some other separator
+value you can pass this with the ``delim`` parameter. You can specify the sourcetype to upload the data to with that ``table_name`` parameter but by default the uploader will upload to the sourcetype with the same name as the file.
```
Comment (ianhelle) [inline docs/source/data_acquisition/UploadData.rst:125]: Trye to keep lines wrapped to about 80-90 chars

**Comment by ianhelle** [comment]
Looks like you need to fix a couple of tests as well because of the parameter change. You can add tests that use both table_name and source_type to check that they both work.

**Comment by Tatsuya-hasegawa** [comment]
Hi @ianhelle 

Thanks for your check. I agree with your comments.

I wanted to rename the parameter of table_name to source_type(sourcetype).
Unfortunately, pylint W0237 prevented it.

pylint...................................................................Failed
msticpy/data/uploaders/splunk_uploader.py:166:4: W0237: Parameter 'table_name' has been renamed to 'sourcetype' in overriding 'SplunkUploader.upload_file' method (arguments-renamed)
msticpy/data/uploaders/splunk_uploader.py:218:4: W0237: Parameter 'table_name' has been renamed to 'sourcetype' in overriding 'SplunkUploader.upload_folder' method (arguments-renamed)

I don't know the way of bypass this check. Would you teach me ?
In addition, regarding to your last comment of 
> You can add tests that use both table_name and source_type to check that they both work.
How do I write that test ? Don't you mind if you give me a sample ?

Sincerely,

**Comment by Tatsuya-hasegawa** [comment]
Dear @ianhelle 

Thanks for looking at my PR.
I have fixed codes to be along your comment, in addition,  added the tests.

Could you check again ?

Best regards,

**Review by ianhelle — APPROVED** [review]

File: msticpy/data/uploaders/splunk_uploader.py [hunk 5]
```diff
@@ -97,27 +99,35 @@ def _post_data(
...
             progress.update(1)
         progress.close()
         if self._debug is True:
-            print("Upload complete")
+            print(f"Upload complete: Splunk sourcetype = {source_types}")
 
     # pylint: disable=arguments-differ
     def upload_df(  # type: ignore
         self,
-        data: pd.DataFrame,
```
Comment (ianhelle) [inline msticpy/data/uploaders/splunk_uploader.py:118]: Would be good not to change the order of existing parameters - this is what is breaking tests.
I always add new parameters to the end of the existing named parameters  - that way it won't break existing cases that are using positional parameters.

Comment (Tatsuya-hasegawa) [reply msticpy/data/uploaders/splunk_uploader.py:118]: Thanks very much for your advice!

**Comment by Tatsuya-hasegawa** [comment]
Dear @ianhelle 

Thank you for teaching me many times. 
I have modified along with your advice.
I much appreciate your advice. 

So, I think this will finally passed for CI workflows.
My local code review is all green and my Splunk instance upload tests were also fine!

```
(base) hacket@hackeTlab msticpy % git commit -m "fix the new argument position and slightly modified index_name param default value"
Check Yaml...........................................(no files to check)Skipped
Check JSON...........................................(no files to check)Skipped
Trim Trailing Whitespace.................................................Passed
black....................................................................Passed
pylint...................................................................Passed
flake8...................................................................Passed
isort (python)...........................................................Passed
pydocstyle...............................................................Passed
check_reqs_all...........................................................Passed
[mod_splunk_uploader_post c40ed2e7] fix the new argument position and slightly modified index_name param default value
 1 file changed, 12 insertions(+), 6 deletions(-)
``` 

Best regards,

**Review by ianhelle — APPROVED** [review]

**Comment by Tatsuya-hasegawa** [comment]
Thanks for your support to this merge.

**Pull request merged**